File: backend/app/domain/reports.py
"""Pydantic schemas for analytics endpoints."""
from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel


Period = Literal["7d", "30d", "90d", "all"]


# ---------- Performance (from Trade) ---------- #

class SymbolPerformance(BaseModel):
    symbol: str
    trade_count: int
    win_count: int
    win_rate: float
    net_pnl: float
    net_r: Optional[float] = None


class PerformanceReport(BaseModel):
    period: Period
    total_trades: int
    open_trades: int
    closed_trades: int
    win_count: int
    loss_count: int
    breakeven_count: int
    win_rate: float
    gross_profit: float
    gross_loss: float
    net_pnl: float
    avg_pnl: float
    profit_factor: Optional[float]
    best_r: Optional[float]
    worst_r: Optional[float]
    avg_r: Optional[float]
    expectancy_r: Optional[float]
    max_drawdown: float
    per_symbol: list[SymbolPerformance]


# ---------- Accuracy (from SignalEvent) ---------- #

class SymbolAccuracy(BaseModel):
    symbol: str
    total: int
    hit_tp: int
    hit_sl: int
    expired: int
    pending: int
    hit_rate: Optional[float]      # tp / (tp + sl), None when neither resolved
    avg_time_to_resolve_s: Optional[float]


class AccuracyReport(BaseModel):
    period: Period
    total_signals: int
    hit_tp: int
    hit_sl: int
    expired: int
    pending: int
    hit_rate: Optional[float]
    buy_hit_rate: Optional[float]
    sell_hit_rate: Optional[float]
    avg_time_to_resolve_s: Optional[float]
    per_symbol: list[SymbolAccuracy]


# ---------- Market stats (live, cross-symbol) ---------- #

class MarketStatRow(BaseModel):
    symbol: str
    asset_class: str
    price: float
    change_24h_pct: float
    atr_14_pct: float
    bias: str
    score: int
    zone_context: str


class MarketStatsReport(BaseModel):
    symbols: list[MarketStatRow]
    top_gainers: list[MarketStatRow]
    top_losers: list[MarketStatRow]
    most_volatile: list[MarketStatRow]
    bias_distribution: dict[str, int]


# ---------- Sessions (time-of-day PnL) ---------- #

class SessionBucket(BaseModel):
    session: Literal["asia", "london", "ny", "off"]
    trade_count: int
    win_count: int
    win_rate: float
    net_pnl: float
    avg_r: Optional[float]


class HourBucket(BaseModel):
    hour: int                      # 0..23 UTC
    trade_count: int
    net_pnl: float
    win_rate: float


class SessionsReport(BaseModel):
    period: Period
    sessions: list[SessionBucket]
    hours: list[HourBucket]        # length 24, UTC

File: frontend/features/analytics/reports.ts
/** API Client for Analytics Reports. */

const API_BASE = process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://127.0.0.1:8000";

export type Period = "7d" | "30d" | "90d" | "all";

export interface SymbolPerformance {
  symbol: string;
  trade_count: number;
  win_count: number;
  win_rate: number;
  net_pnl: number;
  net_r: number | null;
}

export interface PerformanceReport {
  period: Period;
  total_trades: number;
  open_trades: number;
  closed_trades: number;
  win_count: number;
  loss_count: number;
  breakeven_count: number;
  win_rate: number;
  gross_profit: number;
  gross_loss: number;
  net_pnl: number;
  avg_pnl: number;
  profit_factor: number | null;
  best_r: number | null;
  worst_r: number | null;
  avg_r: number | null;
  expectancy_r: number | null;
  max_drawdown: number;
  per_symbol: SymbolPerformance[];
}

export interface SymbolAccuracy {
  symbol: string;
  total: number;
  hit_tp: number;
  hit_sl: number;
  expired: number;
  pending: number;
  hit_rate: number | null;
  avg_time_to_resolve_s: number | null;
}

export interface AccuracyReport {
  period: Period;
  total_signals: number;
  hit_tp: number;
  hit_sl: number;
  expired: number;
  pending: number;
  hit_rate: number | null;
  buy_hit_rate: number | null;
  sell_hit_rate: number | null;
  avg_time_to_resolve_s: number | null;
  per_symbol: SymbolAccuracy[];
}

export interface MarketStatRow {
  symbol: string;
  asset_class: string;
  price: number;
  change_24h_pct: number;
  atr_14_pct: number;
  bias: string;
  score: number;
  zone_context: string;
}

export interface MarketStatsReport {
  symbols: MarketStatRow[];
  top_gainers: MarketStatRow[];
  top_losers: MarketStatRow[];
  most_volatile: MarketStatRow[];
  bias_distribution: Record<string, number>;
}

export interface SessionBucket {
  session: "asia" | "london" | "ny" | "off";
  trade_count: number;
  win_count: number;
  win_rate: number;
  net_pnl: number;
  avg_r: number | null;
}

export interface HourBucket {
  hour: number;
  trade_count: number;
  net_pnl: number;
  win_rate: number;
}

export interface SessionsReport {
  period: Period;
  sessions: SessionBucket[];
  hours: HourBucket[];
}

async function fetchReport<T>(path: string, params?: Record<string, string>): Promise<T> {
  const url = new URL(path, API_BASE);
  if (params) {
    for (const [k, v] of Object.entries(params)) {
      if (v) url.searchParams.set(k, v);
    }
  }
  const res = await fetch(url.toString(), { cache: "no-store" });
  if (!res.ok) {
    throw new Error(`Failed to fetch ${path}: ${res.status}`);
  }
  return res.json() as Promise<T>;
}

export function getPerformance(period: Period): Promise<PerformanceReport> {
  return fetchReport("/analytics/performance", { period });
}

export function getAccuracy(period: Period, symbol?: string): Promise<AccuracyReport> {
  return fetchReport("/analytics/accuracy", { period, symbol: symbol ?? "" });
}

export function getMarketStats(): Promise<MarketStatsReport> {
  return fetchReport("/analytics/market-stats");
}

export function getSessions(period: Period): Promise<SessionsReport> {
  return fetchReport("/analytics/sessions", { period });
}
